File: packages/gavio-py/gavio/exporters/otel.py
"""OpenTelemetry-style runtime span export.

This module intentionally stays dependency-light. It maps Gavio runtime events
to a stable OTel-shaped JSON span model that can be written as JSONL or bridged
to a real OpenTelemetry SDK by the application.
"""

from __future__ import annotations

import hashlib
import json
import threading
from collections import defaultdict
from collections.abc import Callable
from copy import deepcopy
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, TextIO

from .base import GavioRuntimeExporter, metadata_only_event

SpanWriter = Callable[[dict[str, Any]], None]


class OtelSpanExporter(GavioRuntimeExporter):
    """Export runtime events as OpenTelemetry-style span JSON lines.

    Pass ``write`` to receive span dictionaries in-process, ``stream`` in tests,
    or ``path`` in applications. Export is metadata-only by default, matching
    the JSONL runtime exporter privacy boundary.
    """

    def __init__(
        self,
        path: str | Path | None = None,
        *,
        stream: TextIO | None = None,
        write: SpanWriter | None = None,
        service_name: str = "gavio",
        metadata_only: bool = True,
    ) -> None:
        destinations = sum(1 for value in (path, stream, write) if value is not None)
        if destinations == 0:
            raise ValueError("OtelSpanExporter requires path, stream, or write")
        if destinations > 1:
            raise ValueError("pass only one of path, stream, or write")
        self.path = Path(path).expanduser() if path is not None else None
        if self.path is not None:
            self.path.parent.mkdir(parents=True, exist_ok=True)
        self._stream = stream
        self._write = write
        self._metadata_only = metadata_only
        self._mapper = OtelSpanMapper(service_name=service_name)
        self._lock = threading.Lock()

    def export_event(self, event: dict[str, Any]) -> None:
        payload = metadata_only_event(event) if self._metadata_only else deepcopy(event)
        for span in self._mapper.consume(payload):
            self._write_span(span)

    def _write_span(self, span: dict[str, Any]) -> None:
        with self._lock:
            if self._write is not None:
                self._write(span)
                return
            line = json.dumps(span, separators=(",", ":"), sort_keys=True) + "\n"
            if self._stream is not None:
                self._stream.write(line)
                return
            assert self.path is not None
            with self.path.open("a", encoding="utf-8") as f:
                f.write(line)

    def flush(self) -> None:
        if self._stream is not None:
            self._stream.flush()


class OtelSpanMapper:
    """Stateful converter from Inspector/Gavio runtime events to span JSON."""

    def __init__(self, *, service_name: str = "gavio") -> None:
        self.service_name = service_name
        self._traces: dict[str, _TraceState] = {}

    def consume(self, event: dict[str, Any]) -> list[dict[str, Any]]:
        event_type = str(event.get("type", ""))
        trace_id = str(event.get("traceId", ""))
        if not trace_id:
            return []
        if event_type == "trace.start":
            self._traces[trace_id] = _TraceState.from_start(event, self.service_name)
            return []
        state = self._traces.get(trace_id)
        if state is None:
            return []
        if event_type.endswith(".start"):
            state.open_span(event)
            return []
        if event_type == "interceptor.before.end" or event_type == "interceptor.after.end":
            return state.close_interceptor(event)
        if event_type == "provider.call.end":
            return state.close_provider(event)
        if event_type == "trace.error":
            state.add_exception(event)
            return []
        if event_type == "governance.event":
            state.add_event("gavio.governance", event, dict(event.get("data") or {}))
            return []
        if event_type == "trace.end":
            span = state.close_root(event)
            self._traces.pop(trace_id, None)
            return [span]
        return []


def otel_spans_from_events(
    events: list[dict[str, Any]], *, service_name: str = "gavio", metadata_only: bool = True
) -> list[dict[str, Any]]:
    """Map a list of runtime events to OTel-style span JSON."""

    mapper = OtelSpanMapper(service_name=service_name)
    spans: list[dict[str, Any]] = []
    for event in events:
        payload = metadata_only_event(event) if metadata_only else deepcopy(event)
        spans.extend(mapper.consume(payload))
    return spans


class _TraceState:
    def __init__(self, start_event: dict[str, Any], service_name: str) -> None:
        self.start_event = start_event
        self.start_data = dict(start_event.get("data") or {})
        self.service_name = service_name
        self.original_trace_id = str(start_event.get("traceId"))
        self.otel_trace_id = _hex_id(self.original_trace_id, 32)
        self.root_span_id = _hex_id(f"{self.original_trace_id}:root", 16)
        self.root_start_ns = _wall_time_ns(self.start_data.get("wallTimeUtc"))
        self._open: dict[str, list[dict[str, Any]]] = defaultdict(list)
        self._root_events: list[dict[str, Any]] = []

    @classmethod
    def from_start(cls, event: dict[str, Any], service_name: str) -> _TraceState:
        return cls(event, service_name)

    def open_span(self, event: dict[str, Any]) -> None:
        self._open[_open_key(event)].append(event)

    def close_interceptor(self, end_event: dict[str, Any]) -> list[dict[str, Any]]:
        phase = "before" if end_event.get("type") == "interceptor.before.end" else "after"
        data = dict(end_event.get("data") or {})
        start_event = self._pop_open(f"interceptor.{phase}", data.get("name"))
        if start_event is None:
            return []
        name = str(data.get("name", "unknown"))
        attributes = self._base_attributes()
        attributes.update(
            {
                "gavio.interceptor.name": name,
                "gavio.interceptor.phase": phase,
                "gavio.interceptor.mutated": bool(data.get("mutated", False)),
            }
        )
        if "durationUs" in data:
            attributes["gavio.duration_us"] = data["durationUs"]
        decision = data.get("decision")
        if isinstance(decision, dict):
            for key, value in _flatten("gavio.decision", decision).items():
                attributes[key] = value
        return [
            self._span(
                name=f"gavio.interceptor.{phase} {name}",
                logical_key=f"interceptor.{phase}:{name}:{start_event.get('seq')}",
                start_event=start_event,
                end_event=end_event,
                attributes=attributes,
                error=False,
            )
        ]

    def close_provider(self, end_event: dict[str, Any]) -> list[dict[str, Any]]:
        end_data = dict(end_event.get("data") or {})
        start_event = self._pop_open("provider.call", end_data.get("attempt"))
        if start_event is None:
            return []
        start_data = dict(start_event.get("data") or {})
        model = str(start_data.get("model") or self.start_data.get("model") or "unknown")
        provider = str(start_data.get("provider") or self.start_data.get("provider") or "unknown")
        status = str(end_data.get("status") or "ok")
        attributes = self._base_attributes()
        attributes.update(
            {
                "gen_ai.system": provider,
                "gen_ai.request.model": model,
            }
        )
        if "modelVersion" in end_data:
            attributes["gen_ai.response.model"] = end_data["modelVersion"]
        if "attempt" in end_data:
            attributes["gavio.retry.attempt"] = end_data["attempt"]
        usage = end_data.get("usage")
        if isinstance(usage, dict):
            _copy_if_present(attributes, usage, "promptTokens", "gen_ai.usage.input_tokens")
            _copy_if_present(attributes, usage, "completionTokens", "gen_ai.usage.output_tokens")
            _copy_if_present(attributes, usage, "totalTokens", "gen_ai.usage.total_tokens")
        _copy_if_present(attributes, end_data, "costUsd", "gen_ai.usage.cost")
        _copy_if_present(attributes, end_data, "durationUs", "gavio.duration_us")
        _copy_if_present(attributes, end_data, "errorType", "error.type")
        return [
            self._span(
                name=f"chat {model}",
                logical_key=f"provider:{end_data.get('attempt', start_event.get('seq'))}",
                start_event=start_event,
                end_event=end_event,
                attributes=attributes,
                error=status != "ok",
                status_message=str(end_data.get("errorType", "")) or None,
            )
        ]

    def add_exception(self, event: dict[str, Any]) -> None:
        data = dict(event.get("data") or {})
        attrs = {
            "exception.type": data.get("errorType", "Error"),
            "exception.message": data.get("message", ""),
            "gavio.error.origin": data.get("origin", "chain"),
            "exception.escaped": not bool(data.get("handled", False)),
        }
        if "interceptorName" in data:
            attrs["gavio.interceptor.name"] = data["interceptorName"]
        self.add_event("exception", event, attrs)

    def add_event(self, name: str, event: dict[str, Any], attributes: dict[str, Any]) -> None:
        self._root_events.append(
            {
                "name": name,
                "timeUnixNano": self._time_ns(event),
                "attributes": _clean(attributes),
            }
        )

    def close_root(self, end_event: dict[str, Any]) -> dict[str, Any]:
        start_data = self.start_data
        end_data = dict(end_event.get("data") or {})
        attributes = self._base_attributes()
        _copy_if_present(attributes, start_data, "agentId", "gavio.agent_id")
        _copy_if_present(attributes, start_data, "sessionId", "session.id")
        _copy_if_present(attributes, start_data, "parentTraceId", "gavio.parent_trace_id")
        _copy_if_present(attributes, start_data, "provider", "gen_ai.system")
        _copy_if_present(attributes, start_data, "model", "gen_ai.request.model")
        _copy_if_present(attributes, end_data, "latencyMs", "gavio.latency_ms")
        _copy_if_present(attributes, end_data, "costUsd", "gen_ai.usage.cost")
        _copy_if_present(attributes, end_data, "cacheHit", "gavio.cache.hit")
        _copy_if_present(attributes, end_data, "cacheType", "gavio.cache.type")
        _copy_if_present(attributes, end_data, "piiEntityTypes", "gavio.pii.entity_types")
        _copy_if_present(attributes, end_data, "interceptorsFired", "gavio.interceptors")
        dimensions = start_data.get("costDimensions")
        if isinstance(dimensions, dict):
            for key, value in dimensions.items():
                attributes[f"gavio.cost.dimension.{key}"] = value
        return self._span(
            name="gavio.request",
            logical_key="root",
            start_event=self.start_event,
            end_event=end_event,
            attributes=attributes,
            error=str(end_data.get("status", "ok")) != "ok",
            status_message=str(end_data.get("status", "")) or None,
            parent_span_id=None,
            span_id=self.root_span_id,
            events=list(self._root_events),
        )

    def _span(
        self,
        *,
        name: str,
        logical_key: str,
        start_event: dict[str, Any],
        end_event: dict[str, Any],
        attributes: dict[str, Any],
        error: bool,
        status_message: str | None = None,
        parent_span_id: str | None = "root",
        span_id: str | None = None,
        events: list[dict[str, Any]] | None = None,
    ) -> dict[str, Any]:
        resolved_parent = (
            None
            if parent_span_id is None
            else self.root_span_id
            if parent_span_id == "root"
            else parent_span_id
        )
        status: dict[str, Any] = {"code": "ERROR" if error else "OK"}
        if error and status_message:
            status["message"] = status_message
        return {
            "traceId": self.otel_trace_id,
            "spanId": span_id or _hex_id(f"{self.original_trace_id}:{logical_key}", 16),
            "parentSpanId": resolved_parent,
            "name": name,
            "kind": "INTERNAL",
            "startTimeUnixNano": self._time_ns(start_event),
            "endTimeUnixNano": self._time_ns(end_event),
            "status": status,
            "attributes": _clean(attributes),
            "events": events or [],
        }

    def _base_attributes(self) -> dict[str, Any]:
        return {
            "service.name": self.service_name,
            "gavio.trace_id": self.original_trace_id,
            "gavio.event.schema_version": self.start_event.get("schemaVersion", "1.0"),
        }

    def _pop_open(self, family: str, discriminator: Any) -> dict[str, Any] | None:
        key = f"{family}:{discriminator}"
        if self._open.get(key):
            return self._open[key].pop()
        fallback = next(
            (k for k in self._open if k.startswith(f"{family}:") and self._open[k]),
            None,
        )
        if fallback is None:
            return None
        return self._open[fallback].pop()

    def _time_ns(self, event: dict[str, Any]) -> int:
        return self.root_start_ns + int(event.get("tNs") or 0)


def _open_key(event: dict[str, Any]) -> str:
    data = dict(event.get("data") or {})
    event_type = str(event.get("type"))
    if event_type.startswith("interceptor."):
        phase = "before" if event_type == "interceptor.before.start" else "after"
        return f"interceptor.{phase}:{data.get('name')}"
    if event_type == "provider.call.start":
        return f"provider.call:{data.get('attempt')}"
    return f"{event_type}:{event.get('seq')}"


def _wall_time_ns(value: Any) -> int:
    if not isinstance(value, str) or not value:
        return 0
    text = value.replace("Z", "+00:00")
    try:
        dt = datetime.fromisoformat(text)
    except ValueError:
        return 0
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return int(dt.timestamp() * 1_000_000_000)


def _hex_id(seed: str, length: int) -> str:
    return hashlib.sha256(seed.encode("utf-8")).hexdigest()[:length]


def _flatten(prefix: str, value: dict[str, Any]) -> dict[str, Any]:
    out: dict[str, Any] = {}
    for key, raw in value.items():
        name = f"{prefix}.{key}"
        if isinstance(raw, dict):
            out.update(_flatten(name, raw))
        else:
            out[name] = raw
    return out


def _copy_if_present(
    target: dict[str, Any], source: dict[str, Any], source_key: str, target_key: str
) -> None:
    value = source.get(source_key)
    if value is not None:
        target[target_key] = value


def _clean(value: Any) -> Any:
    if isinstance(value, dict):
        return {k: _clean(v) for k, v in value.items() if v is not None}
    if isinstance(value, list):
        return [_clean(item) for item in value]
    return value

File: packages/gavio-js/src/exporters/otel.ts
/** OpenTelemetry-style runtime span export. */

import { appendFileSync, mkdirSync } from 'node:fs'
import { dirname } from 'node:path'
import { createHash } from 'node:crypto'
import type { GavioRuntimeEvent, GavioRuntimeExporter } from './base.js'
import { metadataOnlyEvent } from './base.js'

export interface OtelSpan {
  traceId: string
  spanId: string
  parentSpanId: string | null
  name: string
  kind: 'INTERNAL'
  startTimeUnixNano: number
  endTimeUnixNano: number
  status: { code: 'OK' | 'ERROR' | 'UNSET'; message?: string }
  attributes: Record<string, unknown>
  events: Array<{ name: string; timeUnixNano: number; attributes: Record<string, unknown> }>
}

export interface OtelSpanExporterOptions {
  path?: string
  write?: (span: OtelSpan) => void
  writeLine?: (line: string) => void
  serviceName?: string
  metadataOnly?: boolean
}

export function otelSpanExporter(options: OtelSpanExporterOptions = {}): GavioRuntimeExporter {
  const destinations = [options.path, options.write, options.writeLine].filter(
    (value) => value !== undefined,
  ).length
  if (destinations === 0) throw new Error('otelSpanExporter requires path, write, or writeLine')
  if (destinations > 1) throw new Error('pass only one of path, write, or writeLine')
  if (options.path !== undefined) mkdirSync(dirname(options.path), { recursive: true })
  const mapper = new OtelSpanMapper({ serviceName: options.serviceName ?? 'gavio' })
  const metadataOnly = options.metadataOnly ?? true
  const writeLine =
    options.writeLine ?? ((line: string) => appendFileSync(options.path!, line, 'utf8'))

  return {
    exportEvent(event: GavioRuntimeEvent): void {
      const payload = metadataOnly ? metadataOnlyEvent(event) : structuredClone(event)
      for (const span of mapper.consume(payload)) {
        if (options.write !== undefined) options.write(span)
        else writeLine(JSON.stringify(span) + '\n')
      }
    },
  }
}

export function otelSpansFromEvents(
  events: GavioRuntimeEvent[],
  options: { serviceName?: string; metadataOnly?: boolean } = {},
): OtelSpan[] {
  const mapper = new OtelSpanMapper({ serviceName: options.serviceName ?? 'gavio' })
  const metadataOnly = options.metadataOnly ?? true
  const spans: OtelSpan[] = []
  for (const event of events) {
    const payload = metadataOnly ? metadataOnlyEvent(event) : structuredClone(event)
    spans.push(...mapper.consume(payload))
  }
  return spans
}

class OtelSpanMapper {
  private readonly serviceName: string
  private readonly traces = new Map<string, TraceState>()

  constructor(options: { serviceName: string }) {
    this.serviceName = options.serviceName
  }

  consume(event: GavioRuntimeEvent): OtelSpan[] {
    const traceId = event.traceId
    if (event.type === 'trace.start') {
      this.traces.set(traceId, new TraceState(event, this.serviceName))
      return []
    }
    const state = this.traces.get(traceId)
    if (state === undefined) return []
    if (event.type.endsWith('.start')) {
      state.openSpan(event)
      return []
    }
    if (event.type === 'interceptor.before.end' || event.type === 'interceptor.after.end') {
      return state.closeInterceptor(event)
    }
    if (event.type === 'provider.call.end') return state.closeProvider(event)
    if (event.type === 'trace.error') {
      state.addException(event)
      return []
    }
    if (event.type === 'governance.event') {
      state.addEvent('gavio.governance', event, { ...event.data })
      return []
    }
    if (event.type === 'trace.end') {
      const span = state.closeRoot(event)
      this.traces.delete(traceId)
      return [span]
    }
    return []
  }
}

class TraceState {
  private readonly startEvent: GavioRuntimeEvent
  private readonly serviceName: string
  private readonly originalTraceId: string
  private readonly otelTraceId: string
  private readonly rootSpanId: string
  private readonly rootStartNs: number
  private readonly open = new Map<string, GavioRuntimeEvent[]>()
  private readonly rootEvents: OtelSpan['events'] = []

  constructor(startEvent: GavioRuntimeEvent, serviceName: string) {
    this.startEvent = startEvent
    this.serviceName = serviceName
    this.originalTraceId = startEvent.traceId
    this.otelTraceId = hexId(this.originalTraceId, 32)
    this.rootSpanId = hexId(`${this.originalTraceId}:root`, 16)
    this.rootStartNs = wallTimeNs(asString(startEvent.data['wallTimeUtc']))
  }

  openSpan(event: GavioRuntimeEvent): void {
    const key = openKey(event)
    const list = this.open.get(key) ?? []
    list.push(event)
    this.open.set(key, list)
  }

  closeInterceptor(endEvent: GavioRuntimeEvent): OtelSpan[] {
    const phase = endEvent.type === 'interceptor.before.end' ? 'before' : 'after'
    const name = asString(endEvent.data['name']) ?? 'unknown'
    const startEvent = this.popOpen(`interceptor.${phase}`, name)
    if (startEvent === undefined) return []
    const attributes = this.baseAttributes()
    attributes['gavio.interceptor.name'] = name
    attributes['gavio.interceptor.phase'] = phase
    attributes['gavio.interceptor.mutated'] = Boolean(endEvent.data['mutated'])
    copyIfPresent(attributes, endEvent.data, 'durationUs', 'gavio.duration_us')
    const decision = recordValue(endEvent.data['decision'])
    if (decision !== undefined) {
      Object.assign(attributes, flatten('gavio.decision', decision))
    }
    return [
      this.span({
        name: `gavio.interceptor.${phase} ${name}`,
        logicalKey: `interceptor.${phase}:${name}:${startEvent.seq}`,
        startEvent,
        endEvent,
        attributes,
        error: false,
      }),
    ]
  }

  closeProvider(endEvent: GavioRuntimeEvent): OtelSpan[] {
    const startEvent = this.popOpen('provider.call', endEvent.data['attempt'])
    if (startEvent === undefined) return []
    const model = asString(startEvent.data['model']) ?? asString(this.startEvent.data['model']) ?? 'unknown'
    const provider =
      asString(startEvent.data['provider']) ?? asString(this.startEvent.data['provider']) ?? 'unknown'
    const attributes = this.baseAttributes()
    attributes['gen_ai.system'] = provider
    attributes['gen_ai.request.model'] = model
    copyIfPresent(attributes, endEvent.data, 'modelVersion', 'gen_ai.response.model')
    copyIfPresent(attributes, endEvent.data, 'attempt', 'gavio.retry.attempt')
    copyIfPresent(attributes, endEvent.data, 'costUsd', 'gen_ai.usage.cost')
    copyIfPresent(attributes, endEvent.data, 'durationUs', 'gavio.duration_us')
    copyIfPresent(attributes, endEvent.data, 'errorType', 'error.type')
    const usage = recordValue(endEvent.data['usage'])
    if (usage !== undefined) {
      copyIfPresent(attributes, usage, 'promptTokens', 'gen_ai.usage.input_tokens')
      copyIfPresent(attributes, usage, 'completionTokens', 'gen_ai.usage.output_tokens')
      copyIfPresent(attributes, usage, 'totalTokens', 'gen_ai.usage.total_tokens')
    }
    const status = asString(endEvent.data['status']) ?? 'ok'
    return [
      this.span({
        name: `chat ${model}`,
        logicalKey: `provider:${String(endEvent.data['attempt'] ?? startEvent.seq)}`,
        startEvent,
        endEvent,
        attributes,
        error: status !== 'ok',
        statusMessage: asString(endEvent.data['errorType']),
      }),
    ]
  }

  addException(event: GavioRuntimeEvent): void {
    const attributes: Record<string, unknown> = {
      'exception.type': event.data['errorType'] ?? 'Error',
      'exception.message': event.data['message'] ?? '',
      'gavio.error.origin': event.data['origin'] ?? 'chain',
      'exception.escaped': !Boolean(event.data['handled']),
    }
    copyIfPresent(attributes, event.data, 'interceptorName', 'gavio.interceptor.name')
    this.addEvent('exception', event, attributes)
  }

  addEvent(name: string, event: GavioRuntimeEvent, attributes: Record<string, unknown>): void {
    this.rootEvents.push({
      name,
      timeUnixNano: this.timeNs(event),
      attributes: clean(attributes),
    })
  }

  closeRoot(endEvent: GavioRuntimeEvent): OtelSpan {
    const attributes = this.baseAttributes()
    copyIfPresent(attributes, this.startEvent.data, 'agentId', 'gavio.agent_id')
    copyIfPresent(attributes, this.startEvent.data, 'sessionId', 'session.id')
    copyIfPresent(attributes, this.startEvent.data, 'parentTraceId', 'gavio.parent_trace_id')
    copyIfPresent(attributes, this.startEvent.data, 'provider', 'gen_ai.system')
    copyIfPresent(attributes, this.startEvent.data, 'model', 'gen_ai.request.model')
    copyIfPresent(attributes, endEvent.data, 'latencyMs', 'gavio.latency_ms')
    copyIfPresent(attributes, endEvent.data, 'costUsd', 'gen_ai.usage.cost')
    copyIfPresent(attributes, endEvent.data, 'cacheHit', 'gavio.cache.hit')
    copyIfPresent(attributes, endEvent.data, 'cacheType', 'gavio.cache.type')
    copyIfPresent(attributes, endEvent.data, 'piiEntityTypes', 'gavio.pii.entity_types')
    copyIfPresent(attributes, endEvent.data, 'interceptorsFired', 'gavio.interceptors')
    const dimensions = recordValue(this.startEvent.data['costDimensions'])
    if (dimensions !== undefined) {
      for (const [key, value] of Object.entries(dimensions)) {
        attributes[`gavio.cost.dimension.${key}`] = value
      }
    }
    const status = asString(endEvent.data['status']) ?? 'ok'
    return this.span({
      name: 'gavio.request',
      logicalKey: 'root',
      startEvent: this.startEvent,
      endEvent,
      attributes,
      error: status !== 'ok',
      statusMessage: status,
      parentSpanId: null,
      spanId: this.rootSpanId,
      events: [...this.rootEvents],
    })
  }

  private span(options: {
    name: string
    logicalKey: string
    startEvent: GavioRuntimeEvent
    endEvent: GavioRuntimeEvent
    attributes: Record<string, unknown>
    error: boolean
    statusMessage?: string
    parentSpanId?: string | null
    spanId?: string
    events?: OtelSpan['events']
  }): OtelSpan {
    const status: OtelSpan['status'] = { code: options.error ? 'ERROR' : 'OK' }
    if (options.error && options.statusMessage !== undefined && options.statusMessage !== '') {
      status.message = options.statusMessage
    }
    const parent =
      options.parentSpanId === undefined
        ? this.rootSpanId
        : options.parentSpanId === null
          ? null
          : options.parentSpanId
    return {
      traceId: this.otelTraceId,
      spanId: options.spanId ?? hexId(`${this.originalTraceId}:${options.logicalKey}`, 16),
      parentSpanId: parent,
      name: options.name,
      kind: 'INTERNAL',
      startTimeUnixNano: this.timeNs(options.startEvent),
      endTimeUnixNano: this.timeNs(options.endEvent),
      status,
      attributes: clean(options.attributes),
      events: options.events ?? [],
    }
  }

  private baseAttributes(): Record<string, unknown> {
    return {
      'service.name': this.serviceName,
      'gavio.trace_id': this.originalTraceId,
      'gavio.event.schema_version': this.startEvent.schemaVersion,
    }
  }

  private popOpen(family: string, discriminator: unknown): GavioRuntimeEvent | undefined {
    const exact = `${family}:${String(discriminator)}`
    const exactList = this.open.get(exact)
    if (exactList !== undefined && exactList.length > 0) return exactList.pop()
    for (const [key, list] of this.open) {
      if (key.startsWith(`${family}:`) && list.length > 0) return list.pop()
    }
    return undefined
  }

  private timeNs(event: GavioRuntimeEvent): number {
    return this.rootStartNs + Number(event.tNs ?? 0)
  }
}

function openKey(event: GavioRuntimeEvent): string {
  if (event.type === 'interceptor.before.start') return `interceptor.before:${String(event.data['name'])}`
  if (event.type === 'interceptor.after.start') return `interceptor.after:${String(event.data['name'])}`
  if (event.type === 'provider.call.start') return `provider.call:${String(event.data['attempt'])}`
  return `${event.type}:${event.seq}`
}

function wallTimeNs(value: string | undefined): number {
  if (value === undefined || value === '') return 0
  const ms = Date.parse(value)
  return Number.isFinite(ms) ? ms * 1_000_000 : 0
}

function hexId(seed: string, length: number): string {
  return createHash('sha256').update(seed).digest('hex').slice(0, length)
}

function flatten(prefix: string, value: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {}
  for (const [key, raw] of Object.entries(value)) {
    const name = `${prefix}.${key}`
    const nested = recordValue(raw)
    if (nested !== undefined) Object.assign(out, flatten(name, nested))
    else out[name] = raw
  }
  return out
}

function copyIfPresent(
  target: Record<string, unknown>,
  source: Record<string, unknown>,
  sourceKey: string,
  targetKey: string,
): void {
  const value = source[sourceKey]
  if (value !== undefined && value !== null) target[targetKey] = value
}

function recordValue(value: unknown): Record<string, unknown> | undefined {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) return undefined
  return value as Record<string, unknown>
}

function asString(value: unknown): string | undefined {
  return typeof value === 'string' && value !== '' ? value : undefined
}

function clean(value: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {}
  for (const [key, raw] of Object.entries(value)) {
    if (raw !== undefined && raw !== null) out[key] = raw
  }
  return out
}
